git_of_theseus/analyze.py

```diff
@@ -26,154 +34,226 @@
+            self.last_file_y[path] = file_y
+            processed_entries += 1
+            bar.update()
+
+        return self.cur_y
 
-def analyze(repo, cohortfm='%Y', interval=7*24*60*60, ignore=[], only=[], outdir='.', branch='master', all_filetypes=False, ignore_whitespace=False):
-    repo = git.Repo(repo)
+
+def analyze(repo_dir, cohortfm='%Y', interval=7 * 24 * 60 * 60, ignore=[], only=[], outdir='.', branch='master',
+            all_filetypes=False, ignore_whitespace=False, procs=2, quiet=False, opt=False):
+    repo = git.Repo(repo_dir)
+    blame_kwargs = {}
+    if ignore_whitespace:
+        blame_kwargs['w'] = True
+    master_commits = []  # only stores a subset
     commit2cohort = {}
-    code_commits = [] # only stores a subset
-    master_commits = []
     commit2timestamp = {}
-    curves_set = set()
+    curve_key_tuples = set()  # Keys of each curve that will be tracked
+    tqdm_args = {
+        'smoothing': 0.025,  # Exponential smoothing is still rather jumpy, a tiny number will do
+        'disable': quiet
+    }
     if not os.path.exists(outdir):
         os.makedirs(outdir)
 
+    # Check if specified branch exists
     try:
-        repo.git.checkout(branch)
-    except repo.exc.GitCommandError:
+        repo.git.show_ref('refs/heads/{:s}'.format(branch), verify=True)
+    except git.exc.GitCommandError:
         default_branch = repo.active_branch.name
-        warnings.warn("Requested branch: '{}' does not exist. Falling back to default branch: '{}'".format(branch, default_branch))
+        warnings.warn("Requested branch: '{:s}' does not exist. Falling back to default branch: '{:s}'".format(branch, default_branch))
 
         branch = default_branch
 
-    print('Listing all commits')
-    with progressbar.ProgressBar(max_value=progressbar.UnknownLength, widget_kwargs=widget_kwargs) as bar:
-        for i, commit in enumerate(repo.iter_commits(branch)):
-            bar.update(i)
-            cohort = datetime.datetime.utcfromtimestamp(commit.committed_date).strftime(cohortfm)
-            commit2cohort[commit.hexsha] = cohort
-            curves_set.add(('cohort', cohort))
-            curves_set.add(('author', commit.author.name))
-            if len(commit.parents) == 1:
-                code_commits.append(commit)
-                last_date = commit.committed_date
-                commit2timestamp[commit.hexsha] = commit.committed_date
+    if not quiet and repo.git.version_info < (2, 31, 0):
+        print('Old Git version {:d}.{:d}.{:d} detected. There are optimizations available in version 2.31.0 which speed up performance'.format(*repo.git.version_info))
+
+    if opt:
+        if not quiet:
+            print('Generating git commit-graph... If you wish, this file is deletable later at .git/objects/info')
+        repo.git.execute(['git', 'commit-graph', 'write', '--changed-paths'])  # repo.git.commit_graph('write --changed-paths') doesn't work for some reason
```

Comment:
what's this for?

Reply:
git blame learned to used git commit-graph a while back, it speeds up git blame by 20-40%